src/coreclr/jit/optimizer.cpp

```diff
@@ -4278,40 +4294,65 @@ void Compiler::optInvertWhileLoop(BasicBlock* block)
         }
     }
 
-    // If the compare has too high cost then we don't want to dup
+    // If the compare has too high cost then we don't want to dup.
 
     bool costIsTooHigh = (estDupCostSz > maxDupCostSz);
 
-    int countOfHelpers = 0;
+    OptInvertCountTreeInfoType optInvertTotalInfo = {};
     if (costIsTooHigh)
     {
         // If we already know that the cost is acceptable, then don't waste time walking the tree
-        // counting shared static helpers.
+        // counting things to boost the maximum allowed cost.
         //
         // If the loop condition has a shared static helper, we really want this loop converted
         // as not converting the loop will disable loop hoisting, meaning the shared helper will
         // be executed on every loop iteration.
-        fgWalkTreePre(&condTree, CountSharedStaticHelper, &countOfHelpers);
+        //
+        // If the condition has array.Length operations, also boost, as they are likely to be CSE'd.
 
-        if (countOfHelpers > 0)
+        for (Statement* stmt : bTest->Statements())
         {
-            maxDupCostSz += 24 * min(countOfHelpers, (int)(loopIterations + 1.5));
+            GenTree* tree = stmt->GetRootNode();
+
+            OptInvertCountTreeInfoType optInvertInfo = {};
+            fgWalkTreePre(&tree, Compiler::optInvertCountTreeInfo, &optInvertInfo);
+            optInvertTotalInfo.sharedStaticHelperCount += optInvertInfo.sharedStaticHelperCount;
+            optInvertTotalInfo.arrayLengthCount += optInvertInfo.arrayLengthCount;
 
-            // Is the cost too high now?
-            costIsTooHigh = (estDupCostSz > maxDupCostSz);
+            if ((optInvertInfo.sharedStaticHelperCount > 0) || (optInvertTotalInfo.arrayLengthCount > 0))
```

Comment:
shouldn't this be `(optInvertTotalInfo.sharedStaticHelperCount > 0)`?

Reply:
Oh, the problem is it should be `optInvertInfo.arrayLengthCount > 0` -- not a bug, but an inefficiency. The idea is to avoid the recalculation of max cost unless there would be a change.